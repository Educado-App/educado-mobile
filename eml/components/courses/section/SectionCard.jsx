import React, { useState } from 'react';
import { View, Text, Pressable, StyleSheet, Image, TouchableOpacity } from 'react-native';
import { MaterialCommunityIcons } from "@expo/vector-icons";
import Collapsible from "react-native-collapsible";
import { useNavigation } from '@react-navigation/native'

export default function SectionCard({ sectionNumber, description, imageSrc, completed, total }) {
    const navigation = useNavigation();
    const isComplete = completed === total;
    const [isOpen, setIsOpen] = useState(false);


    const toggleDropdown = () => {
        setIsOpen(!isOpen);
    }

    const handleImagePress = () => {
        navigation.navigate('HomeStack'); // Replace with the name of the target screen
    }

    return (
        <View style={styles.shadowWrapper}>
            <Pressable onPress={toggleDropdown} style={styles.wrapper}>

                <View style={[styles.header, isComplete ? styles.headerComplete : {}]}>

                    <Text style={styles.title}>
                        Section {sectionNumber}
                    </Text>
                    <Text style={styles.completionText}>
                        {completed}/{total} completed
                    </Text>
                    <MaterialCommunityIcons
                        name={isOpen ? "chevron-up" : "chevron-down"}
                        size={25}
                        color="gray"
                    />
                </View>

                <Collapsible collapsed={!isOpen}>

                    <View style={styles.lineBreak} />
                    <Text style={styles.descriptionText}>{description}</Text>
                    <View style={styles.dropdownContent}>
                        <TouchableOpacity style={styles.imageContainer} onPress={handleImagePress}>
                            <Image source={imageSrc ? imageSrc : require('../../../assets/sectionThumbnail.png')} style={styles.image} />
                            <MaterialCommunityIcons name="play-circle-outline" size={50} color="white" style={styles.playIcon} />
                        </TouchableOpacity>
                    </View>
                </Collapsible>

            </Pressable>
        </View>
    );
}

const styles = StyleSheet.create({
    shadowWrapper: {
        backgroundColor: 'transparent',
        margin: 8,
        borderRadius: 10,
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2,
        },
        shadowOpacity: 0.3,
        shadowRadius: 4.65,
        elevation: 8,
    },
    wrapper: {
        backgroundColor: "#fff",
        borderRadius: 10,
        marginBottom: 15,
        marginHorizontal: 18,
        overflow: 'hidden',
    },
    header: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        paddingHorizontal: 25,
        paddingVertical: 15,
    },
    title: {
        fontSize: 16,
        color: 'black',
        flex: 1,
    },
    dropdownContent: {
        flex: 1,
        width: '100%',
    },
    lineBreak: {
        height: 1,
        backgroundColor: '#e0e0e0',
    },
    imageContainer: {
        width: '100%',
        height: 300,
        alignItems: 'center',
        justifyContent: 'center',
        position: 'relative',
    },
    playIcon: {
        position: 'absolute',
        zIndex: 1,
        right: 10,
    },
    image: {
        width: '100%',
        height: 300,
        resizeMode: 'cover',
    },
    headerComplete: {
        backgroundColor: '#87CEEB', // Completion color - change to a green color if needed
    },
    completionText: {
        marginRight: 10,
        color: 'black',
    },
    descriptionText: {
        marginHorizontal: 20,
        marginVertical: 10
    },
});
