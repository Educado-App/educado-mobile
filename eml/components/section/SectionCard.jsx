import React, { useState } from 'react';
import { View, Pressable, Image, TouchableOpacity } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import Text from '../general/Text';
import Collapsible from 'react-native-collapsible';
import { useNavigation } from '@react-navigation/native';
import PropTypes from 'prop-types';


/**
 * A component that displays a section card with collapsible content.
 * @param {Object} section - The section object containing the section data.
 * @param {Object} course - The course object containing the course data.
 * @param {Number} progress - The progress containing the student's progress.
 * @returns {JSX.Element} - The SectionCard component.
 */
export default function SectionCard({ section, course, progress }) {


	const navigation = useNavigation();
	const isComplete = progress === section.components.length;
	const inProgress = 0 < progress && progress < section.components.length;
	const notPossible = progress > section.components.length;
	const [isOpen, setIsOpen] = useState(false);
	const backgroundColor = isComplete ? 'bg-limeGreenDarker' : inProgress ? 'bg-cyanBlue' : notPossible ? 'bg-error' : {};

	/**
     * Toggles the dropdown state.
     */
	const toggleDropdown = () => {
		setIsOpen(!isOpen);
	};


	/**
     * Handles the image press event.
     */
	const handleImagePress = () => {
		navigation.navigate('Components', {
			section: section,
			parsedCourse: course,
		});
	};


	return (
		<View>
			<Pressable testID="collapsible" onPress={toggleDropdown} className="bg-projectWhite rounded-lg shadow-lg shadow-opacity-[0.3] mb-[15] mx-[18] overflow-hidden elevation-[8]">
				<View className={'flex-row items-center justify-between px-[25] py-[15] ' + backgroundColor}>
					<Text className="text-[16px] font-bold text-projectBlack flex-[1]">
						{section.title}
					</Text>
					<Text className="mr-[10] text-projectBlack">
						{/* completed */}
						{progress}/{section.components.length} conclu√≠dos
					</Text>
					<MaterialCommunityIcons
						testID={isOpen ? 'chevron-up' : 'chevron-down'}
						name={isOpen ? 'chevron-up' : 'chevron-down'}
						size={25}
						color="gray"
					/>
				</View>

				<Collapsible collapsed={!isOpen}>
					<View className="h-[1] bg-disable" />
					<Text className="mx-[20] my-[10]">{section.description}</Text>
					<View className="w-[100%]">
						<TouchableOpacity className="w-[100%] h-[300] items-center justify-center relative"
							onPress={handleImagePress}>
							<Image source={require('../../assets/images/sectionThumbnail.png')} className="w-[100%] h-[300] object-cover" />
							<View className="absolute z-[1]">
								<MaterialCommunityIcons name="play-circle-outline" size={100} color="lightblue" />
							</View>
						</TouchableOpacity>
					</View>
				</Collapsible>
			</Pressable>
		</View>
	);
}

SectionCard.propTypes = {
	section: PropTypes.object,
	course: PropTypes.object,
	progress: PropTypes.number,
};