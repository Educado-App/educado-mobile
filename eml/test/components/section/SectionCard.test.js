/**
 * This file contains the test suite for the SectionCard component.
 * It imports React, render, fireEvent and renderer from testing-library/react-native.
 * It also imports the SectionCard component from the components/section directory.
 * The useNavigation hook is mocked using jest.mock.
 * A sample section object is created for testing purposes.
 * The test suite includes tests for rendering the component with provided data,
 * displaying the correct status based on progress, and expanding and collapsing the component.
 * @module SectionCard.test
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import SectionCard from '../../../components/section/SectionCard';
import { mockDataAsyncStorage } from '../../mockData/mockDataAsyncStorage';

// Mock the useNavigation hook outside the describe block
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

let sectionCard;

// Sample section data for testing
const mockData = mockDataAsyncStorage();

beforeEach(async () => {
  sectionCard = renderer.create(<SectionCard section={mockData.section} />);
});

afterAll(() => {
  jest.resetModules();
  jest.restoreAllMocks();
});

describe('<SectionCard />', () => {
  it('Renders correctly with provided data', async () => {
    const { getByText } = render(<SectionCard section={mockData.section} />);

    // Check if title and description are displayed
    expect(getByText(mockData.section.title)).toBeTruthy();
    expect(getByText(mockData.section.description)).toBeTruthy();
  });

  /*it('Displays correct status based on progress', async () => {
    const { queryByText } = render(<SectionCard section={mockData.section} />);

    // Use a regular expression to match the 0/total pattern
    const pattern = new RegExp(`0/${mockData.section.total}`);
    expect(queryByText(pattern)).toBeTruthy();
  });*/
  /**
    * Tests if the SectionCard component renders correctly.
    */
  it('renders SectionCard correctly', async () => {
    expect(await sectionCard.toJSON()).toMatchSnapshot();
  });

  it('should expand and collapse when clicked', async () => {
    /**
         * Represents a sample section.
         * @typedef {Object} Section
         * @property {string} title - The title of the section.
         * @property {string} description - The description of the section.
         * @property {number} total - The total number of items in the section.
         */

    const { getByTestId } = render(<SectionCard section={mockData.section} />);

    const collapsibleButton = getByTestId('collapsible');

    // Ensure Collapsible content is initially hidden
    expect(getByTestId('chevron-down')).toBeTruthy();

    // Click the Collapsible button to expand it
    fireEvent.press(collapsibleButton);

    // Ensure Collapsible content is now visible
    expect(getByTestId('chevron-up')).toBeTruthy();

    // Click the Collapsible button to collapse it
    fireEvent.press(collapsibleButton);

    // Ensure Collapsible content is hidden again
    expect(getByTestId('chevron-down')).toBeTruthy();
  });
});