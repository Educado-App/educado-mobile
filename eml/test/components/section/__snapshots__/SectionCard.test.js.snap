// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SectionCard /> renders SectionCard correctly 1`] = `
<View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "backgroundColor": "#FFFFFF",
        },
        {
          "borderBottomLeftRadius": 8,
          "borderBottomRightRadius": 8,
          "borderTopLeftRadius": 8,
          "borderTopRightRadius": 8,
        },
        {
          "shadowColor": "rgba(0, 0, 0, 0.1)",
          "shadowOffset": {
            "height": 10,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 15,
        },
        {
          "marginBottom": 15,
        },
        {
          "marginLeft": 18,
          "marginRight": 18,
        },
        {
          "overflow": "hidden",
        },
        {
          "elevation": 8,
        },
      ]
    }
    testID="collapsible"
  >
    <View
      style={
        [
          {
            "flexDirection": "row",
          },
          {
            "alignItems": "center",
          },
          {
            "justifyContent": "space-between",
          },
          {
            "paddingLeft": 25,
            "paddingRight": 25,
          },
          {
            "paddingBottom": 15,
            "paddingTop": 15,
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            {
              "color": "#383838",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "fontSize": 16,
              },
              {
                "fontFamily": "Montserrat-Bold",
                "fontWeight": "700",
              },
              {
                "color": "#383838",
              },
              {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ],
          ]
        }
      >
        test section
      </Text>
      <Text
        style={
          [
            {
              "fontFamily": "Montserrat-Regular",
            },
            {
              "color": "#383838",
            },
            {
              "fontSize": 16,
            },
            [
              {
                "marginRight": 10,
              },
              {
                "color": "#383838",
              },
            ],
          ]
        }
      >
        /
        0
         concluídos
      </Text>
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "gray",
              "fontSize": 25,
            },
            undefined,
            {
              "fontFamily": "material-community",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
        testID="chevron-down"
      >
        󰅀
      </Text>
    </View>
    <View
      collapsable={false}
      pointerEvents="none"
      style={
        {
          "height": 0,
          "overflow": "hidden",
        }
      }
    >
      <View
        collapsable={false}
        forwardedRef={[Function]}
        onLayout={[Function]}
        style={{}}
      >
        <View
          style={
            [
              {
                "height": 1,
              },
              {
                "backgroundColor": "#DDD",
              },
            ]
          }
        />
        <Text
          style={
            [
              {
                "fontFamily": "Montserrat-Regular",
              },
              {
                "color": "#383838",
              },
              {
                "fontSize": 16,
              },
              [
                {
                  "marginLeft": 20,
                  "marginRight": 20,
                },
                {
                  "marginBottom": 10,
                  "marginTop": 10,
                },
              ],
            ]
          }
        >
          this is a test section
        </Text>
        <View
          style={
            [
              {
                "width": "100%",
              },
            ]
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "height": 300,
                "justifyContent": "center",
                "opacity": 1,
                "position": "relative",
                "width": "100%",
              }
            }
          >
            <Image
              source={1}
              style={
                [
                  {
                    "width": "100%",
                  },
                  {
                    "height": 300,
                  },
                ]
              }
            />
            <View
              style={
                [
                  {
                    "position": "absolute",
                  },
                  {
                    "zIndex": 1,
                  },
                ]
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "lightblue",
                      "fontSize": 100,
                    },
                    undefined,
                    {
                      "fontFamily": "material-community",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                󰐍
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
